### DISCUSSION TITLE
Remediate CVEs on Keycloak container images

### DISCUSSION THREAD

**Discussion by abstractj** [keycloak/keycloak#12082] [Ideas]
## Summary

Today, we publish the Keycloak container images on Quay.io and DockerHub using GitHub actions. From time to time, we trigger manually the rebuild of those images based on the latest tag to update container dependencies for known vulnerabilities.

## Motivation

Regularly update container images is a security best practice, and necessary to provide to our community the latest bug fixes. The current workflow of rebuilding, overriding and having SHAs purged is not ideal considering [the impact on customers and our community](https://github.com/keycloak/keycloak/issues/11463).

### Current workflow

1. A new CVE is reported via Trivy, Quay Security scanner or Keycloak Security mailing list
2. The rebuild of our container images is manually triggered using the latest tag
3. GitHub actions overrides the existent tags, changing the SHA.

### Proposal 2 (from @andreaTP) - Discussion in progress

The simplest solution we can put in place is to add a tag to the image containing [the date](https://github.com/docker/metadata-action#date-format), those images will not be overwritten and enable SHA pinning for the users requesting it.

### Proposal 3

Something on top of what Andrea suggested. Creating images with the date format suffix makes a lot of sense, at the same time it will generate a lot of noise into the registry, so we need to agree on the frequency. So, we have a few options:

- Option 1

Build those images on the weekly basis, appending the date metadata suffix to those images, and link it to latest.

- Option 2

Introduce a new container tag like `18.0.0-nigthly`. The image would be the outcome from the nightly build of Keycloak 18.0.0 for example, with CVE updates. That means being able to override `18.0.0-nightly` keeping the tag 18.0.0 without any SHA changes.

I'd appreciate any input about it, so we can improve the current process.

### Relevant PRs
- https://github.com/keycloak/keycloak/pull/12110
- https://github.com/keycloak/keycloak/pull/12144

**Comment by abstractj** [comment]
cc @ASzc

**Comment by andreaTP** [comment]
@abstractj thanks for posting this discussion!
I miss seeing how separating the "Keycloak base image" helps on the 3rd point of the current workflow ( `GitHub actions overrides the existent tags, changing the SHA.` ), would you mind elaborating?

I think that old images purging is the outstanding pain point we would need to tackle first on this subject.
The simplest solution we can put in place is to add a tag to the image containing [the date](https://github.com/docker/metadata-action#date-format), those images will not be overwritten and enable SHA pinning for the users requesting it.

**Comment by abstractj** [comment]
@andreaTP the overall idea is to create an extra layer with the [Keycloak base image](https://quay.io/repository/abstractj/keycloak-base?tab=tags) and use it in the [Keycloak Server image](https://quay.io/repository/abstractj/keycloak?tab=tags). In this way, for every new CVE we only have to update the base image, and the tag + SHA will remain unchanged in the Keycloak server image.

**Reply by andreaTP** [reply]
I'm lost 😢, even creating a new/separate base image on a new layer you would need to update the final image for the changes to take effect, as far as I know there is no way to update layers independently.

**Comment by abstractj** [comment]
@andreaTP you are correct considering the immutability of those images :). I completely misunderstood the conversation that I had with @ASzc.

I don't see any other alternative other than agree on the format and frequency of those builds, as you suggested. Another thing that would be nice, is to have a changelog for the changes between those builds.

**Reply by andreaTP** [reply]
A changelog would be great!

**Reply by ASzc** [reply]
Sorry I wasn't clearer Bruno

**Comment by thomasdarimont** [comment]
We wrote a blog post about a similar topic (creating a custom keycloak image with zero cves) a week ago: https://blog.codecentric.de/en/2022/05/keycloak-x-but-secure-without-vulnerable-libraries/

We were able to create a CVE free keycloak image (according to Aquasec and whitesource scanners) based on alpine:
https://github.com/thomasdarimont/keycloak-custom-server

Regarding the current keycloak image:
One problem with the current keycloak image is that the ubi base image itself already seems to contain some vulnerable libraries.
So even if the keycloak distribution would contain no more vulnerable libraries scanners would still complain about the detected cve's in the base image.

**Comment by stianst** [comment]
Re-spinning images should be done when needed, and not on a regular cadence. For RH-SSO images we have tools like FreshMaker that takes care of re-spinning images as needed. I'd suggest talking to @ASzc, @pskopek or @hmlnarik on that. If I'm not fully mistaken here RH-SSO images are re-spun when the UBI images are re-spun.

When the image is re-spun the ideal would be to have an incremental build number suffix (18.0.0-1, 18.0.0-2, etc.) rather than a date. That's slightly harder as we'd need to either manually input the number, or have the workflow do it automatically which would be best.

An additional thing to consider here is should we do a microdnf update ourselves, or fully rely on UBI images to be re-spun as needed. The latter would allow us to release new KC images when there are new UBI images released. I would imagine the team managing the UBI images are responsible when it comes to updating those images when needed.

One last thing to mention as well it's not just about re-spinning whenever there is a CVE. There's such a magnitude of false positives reported by CVE scanners, and in addition many true positives that doesn't affect Keycloak at all.

**Reply by stianst** [reply]
I would suggest the following as a plan of action:

1. Adjust the workflow to add an additional tag with the incremental suffix (so we'd have 18.0, 18.0.0, 18.0.0-1, 18.0.0-2, etc.). That fixes the issue with SHA pinning. We'd still manually re-build the images when needed, which we can identify from the scanning we're doing.
2. Consider a more automated approach. For example re-spin automatically when there's new UBI images.

**Reply by andreaTP** [reply]
For whoever is going to pick up this task, I quickly hacked around the resolution with incremental numbers in bash, so that it can be easily integrated in GH Actions:

```bash
#! /bin/bash
set -euo pipefail

image=$1
tag=$2
original_tag=$tag
c=1

while :
do
  tag="${original_tag}-${c}"
  if DOCKER_CLI_EXPERIMENTAL=enabled docker manifest inspect $image:$tag >/dev/null 2>/dev/null; then
      c=$((c+1))
  else
      echo $tag
      break
  fi
done
```

this script should be called with 2 arguments, the image name (e.g. `quay.io/keycloak/keycloak`) and the version from the tag (e.g. `18.0.0`); it will return to stdout the tag to be used for SHA pinning.

**Reply by andreaTP** [reply]
PR to enable SHA pinning as discussed:
https://github.com/keycloak/keycloak/pull/12110

**Reply by abstractj** [reply]
Thank you, @stianst. I had a chat with @andreaTP, and he agreed about taking over the discussion. That means speaking with @ASzc @pskopek @hmlnarik on how we re-spinning those images to cover the issues mentioned here, and propose how we can improve our processes to fix those CVEs.

**Reply by andreaTP** [reply]
Opened a draft PR to automate the images re-spinning in GH action:
https://github.com/keycloak/keycloak/pull/12144

**Reply by jfrantzius** [reply]
Hi @stianst , from what I understand of [Red Hat Container Image Updates](https://access.redhat.com/articles/2208321) for UBI, it would be great to have that automated approach of respinning when a new UBI base image is available. That way, critical base image fixes would find their way immediately into a Keycloak image, as new UBIs are immediately built when a new version of RHEL is released with critical security fixes. Lower severity security fixes land in new UBI images regularly every six weeks. 

It would be great if Keycloak would reflect this great level of security fixes automatically given with UBI base images!

**Reply by jfrantzius** [reply]
The seemingly corresponding https://github.com/keycloak/keycloak/pull/12144 is closed, but I'm not sure it was merged? I'm asking because the current 18.0.2 image has a critical security issue "RHSA-2022:5683: java-11-openjdk security, bug fix, and enhancement update (Important)" that would be fixed in the underlying UBI image: https://quay.io/repository/keycloak/keycloak/manifest/sha256:47962e880d372dc14273e271afc2869b38a512de17e32f3e2eb66ebde3dba710?tab=vulnerabilities
So it seems that at least 18.0.x images aren't respun when new UBI images are available? Should https://github.com/keycloak/keycloak/pull/12144 apply to both 18.0.x as well as 19.0.x?

**Comment by tumbl3w33d** [comment]
The thread's subject suggests the discussion is about remediation of CVEs, however, it digressed to the pinning aspect for those who appreciate stable tags. I'd like to add my perspective on it, as I could only find what I want in @thomasdarimont 's reply (thanks a lot for that).

Keycloak is one of many softwares we use in our infrastructure(s) and the only chance to manage such a multitude is automation (CVE scanners). Therefore we pull everything through our harbor registry (which comes with trivy by default) and get alerted about CVEs (latest when we try to deploy). From what I've read in some issues/discussions here, you also check with trivy (and other scanners) on a regular basis and you pointed out that a lot of false positives are reported by these vulnerability scanners. Since we're also working in software development, we are well aware of that and we also maintain suppression lists after we analyzed the reports.

Now here's my problem: There are two possibilities (I see) to use keycloak and sleep well at night.
1. build an own image as @thomasdarimont demonstrated
  * that's what we currently do, but that's a really heavy-weight solution and if I imagine I'd need to do that for all our software, that would turn into full-time jobs
  * I'd still prefer doing that instead of option two, because I must expose keycloak to the www for it to make sense and well, I don't need to explain that an IAM is a tasty target for attackers so it's certainly worth the effort
  * my point, though – if it's possible to build a "clean" image (and the proof that it's possible is provided), why wouldn't you just do it and completely avoid the entire discussion?
2. I trust the keycloak maintainers blindly that they review vulnerabilities as soon as possible (for me that means daily)
* I'd leave out the commercial subscription aspect here, because if that would change anything, I'd just subscribe to the RH version and be done with it
* From my experience, everything that's not automated will be forgotten now and then (e.g. 'the one guy who always does it' is on parental/sick leave). Unfortunately, reviewing vulnerability impact usually takes in-depth knowledge of the software to make a good judgement. That's why we opt for blindly upgrading dependencies whenever possibly and only go for reviews when there's no fix available.
* A current example: today my trivy shows that keycloak 19.0.2 is vulnerable to [CVE-2022-40153](https://avd.aquasec.com/nvd/cve-2022-40153). It's rated medium/high, depending on the one rating it and is 12d old. I cannot find this CVE number in the github issues and I cannot find info about it in the security section of the github repo. I feel quite helpless in these situations and can only hope that the keycloak maintainers know about it and consider it irrelevant. At the same time, I don't wanna show up in the issues weekly (that's the perceived frequency) and report it, because I've read the maintainers' opinion about 'all these false positives' (and I get their point). But I'd like advise what you would do if you were in my shoes and wouldn't want to blindly rely on a third party.

About the pinning aspect – I understand that some folks prefer stability over security, but I am not one of those. I'd rather shut the entire thing down with no return until there's at least a workaround. I'm not saying there shouldn't be this pinning option for those who like it, but I personally wouldn't use that for the reasons mentioned above. If there's no other option than manually selecting the latest patch version on every update I deploy, I can live with it, though.

**Comment by kforner** [comment]
We are in a similar position as @tumbl3w33d.

> my point, though – if it's possible to build a "clean" image (and the proof that it's possible is provided), why wouldn't you just do it and completely avoid the entire discussion?

Exactly. I understand that it is a lot of work. Should we unite our development efforts for this?!
I would be of course tremendously better if this 0-vulnerability was under the umbrella of the keycloak project thought. 
So what could we do realistically  to achieve that ? 
Could we contribute directly to keycloak on that aspect or should we create a parallel project to build those fixed docker images ?
